File: step_definitions/Simulate failed shipping quote retrieval and verify error handling.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import LoginPage from "../pages/loginPage";
import HomePage from "../pages/homePage";
import CartPage from "../pages/cartPage";
import CheckoutPage from "../pages/checkoutPage";
import ErrorPage from "../pages/errorPage";
import OrderHistoryPage from "../pages/orderHistoryPage";

Given("User account exists", () => {
 // Assume user account exists
});

When("User logs into the application using valid credentials", () => {
 LoginPage.visit();
 LoginPage.enterCredentials('validUser', 'validPassword');
 LoginPage.submit();
});

Then("User is successfully logged in and navigated to the homepage", () => {
 HomePage.verifyHomePage();
});

Given("User is on the homepage", () => {
 HomePage.verifyHomePage();
});

When("User navigates to the shopping cart", () => {
 HomePage.goToCart();
});

Then("All items added to the cart are displayed with their respective details", () => {
 CartPage.verifyItemsInCart();
});

Given("User is on the shopping cart page", () => {
 CartPage.verifyCartPage();
});

When("User proceeds to checkout", () => {
 CartPage.proceedToCheckout();
});

Then("Checkout page is displayed with options to enter shipping information", () => {
 CheckoutPage.verifyCheckoutPage();
});

Given("User is on the checkout page", () => {
 CheckoutPage.verifyCheckoutPage();
});

When("User enters valid shipping address details", () => {
 CheckoutPage.enterShippingDetails('John Doe', '123 Elm St', 'New York', 'NY', '10001');
});

Then("Shipping address is accepted and saved", () => {
 CheckoutPage.verifyShippingDetailsSaved();
});

Given("User has entered shipping address details", () => {
 CheckoutPage.verifyShippingDetailsSaved();
});

When("User selects 'Get Shipping Quotes' option", () => {
 CheckoutPage.getShippingQuotes();
});

Then("System attempts to retrieve shipping quotes", () => {
 CheckoutPage.verifyQuotesRetrievalAttempt();
});

Given("System is attempting to retrieve shipping quotes", () => {
 CheckoutPage.verifyQuotesRetrievalAttempt();
});

When("Network failure or service outage is simulated", () => {
 cy.intercept('GET', '/api/shipping-quotes', { forceNetworkError: true }).as('getQuotes');
 cy.wait('@getQuotes');
});

Then("System displays an error message indicating failure to retrieve shipping quotes", () => {
 ErrorPage.verifyErrorMessageDisplayed();
});

Given("Error message is displayed", () => {
 ErrorPage.verifyErrorMessageDisplayed();
});

Then("Error message includes options to retry or contact support", () => {
 ErrorPage.verifyRetryOrContactOptions();
});

Given("Error message includes retry option", () => {
 ErrorPage.verifyRetryOrContactOptions();
});

When("User attempts to retrieve shipping quotes again using retry option", () => {
 ErrorPage.retryGettingQuotes();
});

Then("System attempts to retrieve quotes again and either succeeds or displays error message", () => {
 CheckoutPage.verifyQuotesRetrievalAttempt();
});

Given("System has displayed an error message", () => {
 ErrorPage.verifyErrorMessageDisplayed();
});

Then("Error details are logged in the system for troubleshooting", () => {
 // Assume error details are logged
});

Given("Error details are logged in the system", () => {
 // Assume error details are logged
});

When("User checks email for notification", () => {
 // Assume user checks email
});

Then("Email notification is sent with details about the issue and possible resolutions", () => {
 // Assume email notification is sent
});

Given("User is on any page of the application", () => {
 // Assume user is on any page
});

When("User logs out from the application", () => {
 HomePage.logout();
});

Then("User is successfully logged out", () => {
 LoginPage.verifyLoggedOut();
});

Given("User is logged out", () => {
 LoginPage.verifyLoggedOut();
});

When("User logs into the application again", () => {
 LoginPage.enterCredentials('validUser', 'validPassword');
 LoginPage.submit();
});

And("User navigates to order history", () => {
 HomePage.goToOrderHistory();
});

Then("Order history displays recent order with an indication of shipping quote retrieval issue", () => {
 OrderHistoryPage.verifyOrderWithQuoteIssue();
});

Given("User is on the order history page", () => {
 OrderHistoryPage.verifyOrderHistoryPage();
});

When("User attempts to edit shipping details", () => {
 OrderHistoryPage.editShippingDetails();
});

And("User attempts to retrieve quotes again", () => {
 OrderHistoryPage.retryGettingQuotes();
});

Then("System allows editing of shipping details and attempts to retrieve quotes", () => {
 CheckoutPage.verifyQuotesRetrievalAttempt();
});

Given("User has edited shipping details", () => {
 // Assume user has edited shipping details
});

When("System successfully retrieves shipping quotes", () => {
 // Assume quotes are successfully retrieved
});

Then("System allows selection of preferred provider", () => {
 CheckoutPage.selectPreferredProvider();
});

Given("User has selected preferred shipping provider", () => {
 // Assume user has selected a provider
});

When("User confirms the shipping selection", () => {
 CheckoutPage.confirmShippingSelection();
});

Then("User is navigated to the payment section with selected shipping details", () => {
 CheckoutPage.verifyNavigationToPayment();
});

File: step_definitions/Simulate failed shipping quote retrieval and verify error handling.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import com.page_objects.ShippingPage;

public class ShippingSteps extends ShippingPage {

    private TestHarness testHarness = new TestHarness();

    @Given("User account exists")
    public void userAccountExists() {
        verifyUserAccountExists();
    }

    @When("User logs into the application using valid credentials")
    public void userLogsIntoApplication() {
        String username = testHarness.getData("LoginData", "Username");
        String password = testHarness.getData("LoginData", "Password");
        login(username, password);
    }

    @Then("User is successfully logged in and navigated to the homepage")
    public void userIsLoggedIn() {
        verifyUserIsOnHomepage();
    }

    @Given("User is on the homepage")
    public void userIsOnHomepage() {
        verifyUserIsOnHomepage();
    }

    @When("User navigates to the shopping cart")
    public void userNavigatesToShoppingCart() {
        navigateToShoppingCart();
    }

    @Then("All items added to the cart are displayed with their respective details")
    public void verifyItemsInCart() {
        verifyCartItems();
    }

    @Given("User is on the shopping cart page")
    public void userIsOnShoppingCartPage() {
        verifyUserIsOnShoppingCartPage();
    }

    @When("User proceeds to checkout")
    public void userProceedsToCheckout() {
        proceedToCheckout();
    }

    @Then("Checkout page is displayed with options to enter shipping information")
    public void verifyCheckoutPage() {
        verifyCheckoutPageDisplayed();
    }

    @Given("User is on the checkout page")
    public void userIsOnCheckoutPage() {
        verifyUserIsOnCheckoutPage();
    }

    @When("User enters valid shipping address details")
    public void userEntersShippingDetails() {
        String fullName = testHarness.getData("ShippingData", "FullName");
        String address = testHarness.getData("ShippingData", "Address");
        String city = testHarness.getData("ShippingData", "City");
        String postalCode = testHarness.getData("ShippingData", "PostalCode");
        String phone = testHarness.getData("ShippingData", "Phone");
        enterShippingDetails(fullName, address, city, postalCode, phone);
    }

    @Then("Shipping address is accepted and saved")
    public void verifyShippingAddressSaved() {
        verifyShippingAddressAccepted();
    }

    @Given("User has entered shipping address details")
    public void userHasEnteredShippingDetails() {
        verifyShippingAddressEntered();
    }

    @When("User selects 'Get Shipping Quotes' option")
    public void userSelectsGetShippingQuotes() {
        selectGetShippingQuotes();
    }

    @Then("System attempts to retrieve shipping quotes")
    public void systemAttemptsToRetrieveQuotes() {
        verifyShippingQuotesRetrieval();
    }

    @Given("System is attempting to retrieve shipping quotes")
    public void systemIsRetrievingQuotes() {
        verifyShippingQuotesRetrieval();
    }

    @When("Network failure or service outage is simulated")
    public void simulateNetworkFailure() {
        simulateNetworkFailureOrOutage();
    }

    @Then("System displays an error message indicating failure to retrieve shipping quotes")
    public void verifyErrorMessageDisplayed() {
        verifyErrorMessageForShippingQuotes();
    }

    @Given("Error message is displayed")
    public void errorMessageIsDisplayed() {
        verifyErrorMessageForShippingQuotes();
    }

    @Then("Error message includes options to retry or contact support")
    public void verifyErrorMessageOptions() {
        verifyErrorMessageOptionsAvailable();
    }

    @Given("Error message includes retry option")
    public void errorMessageIncludesRetryOption() {
        verifyErrorMessageOptionsAvailable();
    }

    @When("User attempts to retrieve shipping quotes again using retry option")
    public void userRetriesShippingQuotes() {
        retryShippingQuotesRetrieval();
    }

    @Then("System attempts to retrieve quotes again and either succeeds or displays error message")
    public void verifyRetryOutcome() {
        verifyRetryOutcomeForShippingQuotes();
    }

    @Given("System has displayed an error message")
    public void systemDisplayedErrorMessage() {
        verifyErrorMessageForShippingQuotes();
    }

    @Then("Error details are logged in the system for troubleshooting")
    public void verifyErrorLogging() {
        verifyErrorDetailsLogged();
    }

    @Given("Error details are logged in the system")
    public void errorDetailsLogged() {
        verifyErrorDetailsLogged();
    }

    @When("User checks email for notification")
    public void userChecksEmail() {
        checkEmailForNotification();
    }

    @Then("Email notification is sent with details about the issue and possible resolutions")
    public void verifyEmailNotification() {
        verifyEmailNotificationSent();
    }

    @Given("User is on any page of the application")
    public void userIsOnAnyPage() {
        verifyUserIsOnAnyPage();
    }

    @When("User logs out from the application")
    public void userLogsOut() {
        logout();
    }

    @Then("User is successfully logged out")
    public void verifyUserLoggedOut() {
        verifyUserIsLoggedOut();
    }

    @Given("User is logged out")
    public void userIsLoggedOut() {
        verifyUserIsLoggedOut();
    }

    @When("User logs into the application again")
    public void userLogsInAgain() {
        String username = testHarness.getData("LoginData", "Username");
        String password = testHarness.getData("LoginData", "Password");
        login(username, password);
    }

    @When("User navigates to order history")
    public void userNavigatesToOrderHistory() {
        navigateToOrderHistory();
    }

    @Then("Order history displays recent order with an indication of shipping quote retrieval issue")
    public void verifyOrderHistory() {
        verifyOrderHistoryForShippingIssue();
    }

    @Given("User is on the order history page")
    public void userIsOnOrderHistoryPage() {
        verifyUserIsOnOrderHistoryPage();
    }

    @When("User attempts to edit shipping details")
    public void userEditsShippingDetails() {
        editShippingDetails();
    }

    @When("User attempts to retrieve quotes again")
    public void userRetrievesQuotesAgain() {
        retryShippingQuotesRetrieval();
    }

    @Then("System allows editing of shipping details and attempts to retrieve quotes")
    public void verifyEditAndRetrieveQuotes() {
        verifyEditAndRetrieveQuotesFunctionality();
    }

    @Given("User has edited shipping details")
    public void userHasEditedShippingDetails() {
        verifyShippingDetailsEdited();
    }

    @When("System successfully retrieves shipping quotes")
    public void systemRetrievesQuotesSuccessfully() {
        verifySuccessfulShippingQuotesRetrieval();
    }

    @Then("System allows selection of preferred provider")
    public void verifyProviderSelection() {
        verifyProviderSelectionAvailable();
    }

    @Given("User has selected preferred shipping provider")
    public void userHasSelectedProvider() {
        verifyProviderSelected();
    }

    @When("User confirms the shipping selection")
    public void userConfirmsShippingSelection() {
        confirmShippingSelection();
    }

    @Then("User is navigated to the payment section with selected shipping details")
    public void verifyNavigationToPayment() {
        verifyNavigationToPaymentSection();
    }
}